# Youn Hee is going to work in here :)

Import stuff

In [1]:
import pandas as pd
import numpy as np
import requests

Data from https://collegescorecard.ed.gov/data/

In [2]:
# load the data from csv
big = pd.read_csv('/Users/yoni/documents/GA_prod/project-5/CollegeScorecard_Raw_Data/MERGED2015_16_PP.csv')

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (6,9,1725,1726,1727,1728,1729,1743) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
ref = pd.read_csv('./college_data_variables.csv')
ref['VARIABLE NAME'].unique()

array(['PCIP01', 'PCIP03', 'PCIP04', 'PCIP05', 'PCIP09', 'PCIP10',
       'PCIP11', 'PCIP12', 'PCIP13', 'PCIP14', 'PCIP15', 'PCIP16',
       'PCIP19', 'PCIP22', 'PCIP23', 'PCIP24', 'PCIP25', 'PCIP26',
       'PCIP27', 'PCIP29', 'PCIP30', 'PCIP31', 'PCIP38', 'PCIP39',
       'PCIP40', 'PCIP41', 'PCIP42', 'PCIP43', 'PCIP44', 'PCIP45',
       'PCIP46', 'PCIP47', 'PCIP48', 'PCIP49', 'PCIP50', 'PCIP51',
       'PCIP52', 'PCIP54', 'CIP01BACHL', 'CIP03BACHL', 'CIP04BACHL',
       'CIP05BACHL', 'CIP09BACHL', 'CIP10BACHL', 'CIP11BACHL',
       'CIP12BACHL', 'CIP13BACHL', 'CIP14BACHL', 'CIP15BACHL',
       'CIP16BACHL', 'CIP19BACHL', 'CIP22BACHL', 'CIP23BACHL',
       'CIP24BACHL', 'CIP25BACHL', 'CIP26BACHL', 'CIP27BACHL',
       'CIP29BACHL', 'CIP30BACHL', 'CIP31BACHL', 'CIP38BACHL',
       'CIP39BACHL', 'CIP40BACHL', 'CIP41BACHL', 'CIP42BACHL',
       'CIP43BACHL', 'CIP44BACHL', 'CIP45BACHL', 'CIP46BACHL',
       'CIP47BACHL', 'CIP48BACHL', 'CIP49BACHL', 'CIP50BACHL',
       'CIP51BACHL', 'CIP

In [4]:
target = big[['PCIP01', 'PCIP03', 'PCIP04', 'PCIP05', 'PCIP09', 'PCIP10',
       'PCIP11', 'PCIP12', 'PCIP13', 'PCIP14', 'PCIP15', 'PCIP16',
       'PCIP19', 'PCIP22', 'PCIP23', 'PCIP24', 'PCIP25', 'PCIP26',
       'PCIP27', 'PCIP29', 'PCIP30', 'PCIP31', 'PCIP38', 'PCIP39',
       'PCIP40', 'PCIP41', 'PCIP42', 'PCIP43', 'PCIP44', 'PCIP45',
       'PCIP46', 'PCIP47', 'PCIP48', 'PCIP49', 'PCIP50', 'PCIP51',
       'PCIP52', 'PCIP54', 'CIP01BACHL', 'CIP03BACHL', 'CIP04BACHL',
       'CIP05BACHL', 'CIP09BACHL', 'CIP10BACHL', 'CIP11BACHL',
       'CIP12BACHL', 'CIP13BACHL', 'CIP14BACHL', 'CIP15BACHL',
       'CIP16BACHL', 'CIP19BACHL', 'CIP22BACHL', 'CIP23BACHL',
       'CIP24BACHL', 'CIP25BACHL', 'CIP26BACHL', 'CIP27BACHL',
       'CIP29BACHL', 'CIP30BACHL', 'CIP31BACHL', 'CIP38BACHL',
       'CIP39BACHL', 'CIP40BACHL', 'CIP41BACHL', 'CIP42BACHL',
       'CIP43BACHL', 'CIP44BACHL', 'CIP45BACHL', 'CIP46BACHL',
       'CIP47BACHL', 'CIP48BACHL', 'CIP49BACHL', 'CIP50BACHL',
       'CIP51BACHL', 'CIP52BACHL', 'CIP54BACHL', 'ADM_RATE', 'SATVRMID',
       'SATMTMID', 'SATWRMID', 'ACTCMMID', 'ACTENMID', 'ACTMTMID',
       'ACTWRMID', 'PCTFLOAN', 'GRAD_DEBT_MDN', 'WDRAW_DEBT_MDN',
       'LO_INC_DEBT_MDN', 'MD_INC_DEBT_MDN', 'HI_INC_DEBT_MDN', 'C150_4',
       'TRANS_4', 'COSTT4_A', 'COSTT4_P', 'TUITIONFEE_IN',
       'TUITIONFEE_OUT', 'MD_EARN_WNE_P6', 'CONTROL', 'REGION', 'LOCALE',
       'LOCALE2', 'TUITFTE', 'INEXPFTE', 'AVGFACSAL', 'PFTFAC', 'UG',
       'UG25ABV', 'PAR_ED_PCT_MS', 'PAR_ED_PCT_HS', 'PAR_ED_PCT_PS',
       'AGE_ENTRY', 'FEMALE', 'D_PCTPELL_PCTFLOAN']]

In [5]:
target.shape

(7593, 113)

In [6]:
target.head()

PCIP01  PCIP03  PCIP04  PCIP05  PCIP09  PCIP10  PCIP11  PCIP12  PCIP13  \
0  0.0446  0.0023  0.0094  0.0000  0.0000  0.0164  0.0634     0.0  0.1268   
1  0.0000  0.0000  0.0000  0.0009  0.0426  0.0000  0.0133     0.0  0.0815   
2  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000     0.0  0.0000   
3  0.0000  0.0000  0.0000  0.0000  0.0216  0.0000  0.0315     0.0  0.0216   
4  0.0000  0.0000  0.0000  0.0000  0.0945  0.0000  0.0567     0.0  0.1531   

   PCIP14         ...          AVGFACSAL  PFTFAC  UG  UG25ABV  PAR_ED_PCT_MS  \
0  0.1432         ...             7017.0  0.7096 NaN   0.0877   0.0192989366   
1  0.0577         ...            10221.0  0.9081 NaN   0.2363   0.0230535015   
2  0.0000         ...             3217.0  1.0000 NaN   0.8571   0.0619047619   
3  0.3027         ...             9514.0  0.6173 NaN   0.2255   0.0275862069   
4  0.0000         ...             7940.0  0.6395 NaN   0.0974   0.0159163256   

   PAR_ED_PCT_HS  PAR_ED_PCT_PS     AGE_ENTRY        FEMALE  \
0   0.3694367861   0.6112642773  20.277133825  0.5472427116   
1   0.3334058286   0.6435406699   23.58415642  0.6369683532   
2   0.4666666667   0.4714285714  33.065934066  0.6043956044   
3   0.2862068966   0.6862068966  23.152713891  0.4779208832   
4   0.3533424284    0.630741246  20.201285657   0.612294094   

   D_PCTPELL_PCTFLOAN  
0              4210.0  
1             11679.0  
2               279.0  
3              5618.0  
4              4805.0  

[5 rows x 113 columns]

In [7]:
target.isnull().sum()

PCIP01                 739
PCIP03                 739
PCIP04                 739
PCIP05                 739
PCIP09                 739
PCIP10                 739
PCIP11                 739
PCIP12                 739
PCIP13                 739
PCIP14                 739
PCIP15                 739
PCIP16                 739
PCIP19                 739
PCIP22                 739
PCIP23                 739
PCIP24                 739
PCIP25                 739
PCIP26                 739
PCIP27                 739
PCIP29                 739
PCIP30                 739
PCIP31                 739
PCIP38                 739
PCIP39                 739
PCIP40                 739
PCIP41                 739
PCIP42                 739
PCIP43                 739
PCIP44                 739
PCIP45                 739
                      ... 
ACTWRMID              7268
PCTFLOAN               758
GRAD_DEBT_MDN           29
WDRAW_DEBT_MDN          29
LO_INC_DEBT_MDN         29
MD_INC_DEBT_MDN         29
H

In [15]:
# columns that Charles is not looking at for the moment
other = target[['ADM_RATE', 'SATVRMID', 'SATMTMID', 'SATWRMID', 'ACTCMMID', 'ACTENMID', 'ACTMTMID','ACTWRMID', 'PCTFLOAN', 
'GRAD_DEBT_MDN', 'WDRAW_DEBT_MDN', 'LO_INC_DEBT_MDN', 'MD_INC_DEBT_MDN', 'HI_INC_DEBT_MDN', 'C150_4',
'TRANS_4', 'COSTT4_A', 'COSTT4_P', 'TUITIONFEE_IN','TUITIONFEE_OUT', 'MD_EARN_WNE_P6', 'CONTROL', 'REGION', 'LOCALE',
'LOCALE2', 'TUITFTE', 'INEXPFTE', 'AVGFACSAL', 'PFTFAC', 'UG', 'UG25ABV', 'PAR_ED_PCT_MS', 'PAR_ED_PCT_HS', 
'PAR_ED_PCT_PS','AGE_ENTRY', 'FEMALE', 'D_PCTPELL_PCTFLOAN']]

In [18]:
other.shape

(7593, 37)

In [16]:
other.head()

ADM_RATE  SATVRMID  SATMTMID  SATWRMID  ACTCMMID  ACTENMID  ACTMTMID  \
0    0.6538     427.0     420.0     414.0      18.0      17.0      17.0   
1    0.6043     575.0     594.0       NaN      25.0      26.0      23.0   
2       NaN       NaN       NaN       NaN       NaN       NaN       NaN   
3    0.8120     585.0     615.0       NaN      27.0      28.0      27.0   
4    0.4639     410.0     410.0       NaN      18.0      17.0      16.0   

   ACTWRMID  PCTFLOAN GRAD_DEBT_MDN         ...         AVGFACSAL  PFTFAC  UG  \
0       NaN    0.8159         35000         ...            7017.0  0.7096 NaN   
1       NaN    0.5218         21500         ...           10221.0  0.9081 NaN   
2       NaN    0.8781         23000         ...            3217.0  1.0000 NaN   
3       NaN    0.4589         23500         ...            9514.0  0.6173 NaN   
4       NaN    0.7692         32091         ...            7940.0  0.6395 NaN   

  UG25ABV  PAR_ED_PCT_MS  PAR_ED_PCT_HS  PAR_ED_PCT_PS     AGE_ENTRY  \
0  0.0877   0.0192989366   0.3694367861   0.6112642773  20.277133825   
1  0.2363   0.0230535015   0.3334058286   0.6435406699   23.58415642   
2  0.8571   0.0619047619   0.4666666667   0.4714285714  33.065934066   
3  0.2255   0.0275862069   0.2862068966   0.6862068966  23.152713891   
4  0.0974   0.0159163256   0.3533424284    0.630741246  20.201285657   

         FEMALE  D_PCTPELL_PCTFLOAN  
0  0.5472427116              4210.0  
1  0.6369683532             11679.0  
2  0.6043956044               279.0  
3  0.4779208832              5618.0  
4   0.612294094              4805.0  

[5 rows x 37 columns]

In [17]:
other.isnull().sum()

ADM_RATE              5422
SATVRMID              6400
SATMTMID              6391
SATWRMID              6860
ACTCMMID              6359
ACTENMID              6500
ACTMTMID              6499
ACTWRMID              7268
PCTFLOAN               758
GRAD_DEBT_MDN           29
WDRAW_DEBT_MDN          29
LO_INC_DEBT_MDN         29
MD_INC_DEBT_MDN         29
HI_INC_DEBT_MDN         29
C150_4                5104
TRANS_4               5104
COSTT4_A              3673
COSTT4_P              5093
TUITIONFEE_IN         3290
TUITIONFEE_OUT        3504
MD_EARN_WNE_P6        7593
CONTROL                  0
REGION                   0
LOCALE                 446
LOCALE2               7593
TUITFTE                460
INEXPFTE               460
AVGFACSAL             3139
PFTFAC                3656
UG                    7593
UG25ABV                805
PAR_ED_PCT_MS          218
PAR_ED_PCT_HS          218
PAR_ED_PCT_PS          218
AGE_ENTRY              218
FEMALE                 218
D_PCTPELL_PCTFLOAN     758
d

In [20]:
other.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7593 entries, 0 to 7592
Data columns (total 37 columns):
ADM_RATE              2171 non-null float64
SATVRMID              1193 non-null float64
SATMTMID              1202 non-null float64
SATWRMID              733 non-null float64
ACTCMMID              1234 non-null float64
ACTENMID              1093 non-null float64
ACTMTMID              1094 non-null float64
ACTWRMID              325 non-null float64
PCTFLOAN              6835 non-null float64
GRAD_DEBT_MDN         7564 non-null object
WDRAW_DEBT_MDN        7564 non-null object
LO_INC_DEBT_MDN       7564 non-null object
MD_INC_DEBT_MDN       7564 non-null object
HI_INC_DEBT_MDN       7564 non-null object
C150_4                2489 non-null float64
TRANS_4               2489 non-null float64
COSTT4_A              3920 non-null float64
COSTT4_P              2500 non-null float64
TUITIONFEE_IN         4303 non-null float64
TUITIONFEE_OUT        4089 non-null float64
MD_EARN_WNE_P6      

#### I suggest that we start with drop following columns, since they are 100% missing values:
- MD_EARN_WNE_P6 (Median earnings of students working and not enrolled 6 years after entry)
- LOCALE2 (Degree of urbanization of institution)
- UG (Enrollment of all undergraduate students) - It says in the documentation that this is only available in 2000-2001

In [22]:
other = other.drop(['MD_EARN_WNE_P6', 'LOCALE2', 'UG'], axis=1)

In [23]:
other.shape

(7593, 34)

In [24]:
other.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7593 entries, 0 to 7592
Data columns (total 34 columns):
ADM_RATE              2171 non-null float64
SATVRMID              1193 non-null float64
SATMTMID              1202 non-null float64
SATWRMID              733 non-null float64
ACTCMMID              1234 non-null float64
ACTENMID              1093 non-null float64
ACTMTMID              1094 non-null float64
ACTWRMID              325 non-null float64
PCTFLOAN              6835 non-null float64
GRAD_DEBT_MDN         7564 non-null object
WDRAW_DEBT_MDN        7564 non-null object
LO_INC_DEBT_MDN       7564 non-null object
MD_INC_DEBT_MDN       7564 non-null object
HI_INC_DEBT_MDN       7564 non-null object
C150_4                2489 non-null float64
TRANS_4               2489 non-null float64
COSTT4_A              3920 non-null float64
COSTT4_P              2500 non-null float64
TUITIONFEE_IN         4303 non-null float64
TUITIONFEE_OUT        4089 non-null float64
CONTROL             

### FACTS ABOUT THE COLUMNS

**TEST SCORES**
- ACTCMMID = Midpoint of the ACT cumulative score<BR /> 
(6359 NULL out of 7593 rows)
- ACTENMID = Midpoint of the ACT English score<BR />
(6500 NULL out of 7593 rows)
- ACTMTMID = Midpoint of the ACT math score<BR /> 
(6490 NULL out of 7593 rows)
- ACTWRMID = Midpoint of the ACT writing score<BR /> 
(7268 NULL out of 7593 rows)
- SATMTMID = Midpoint of SAT scores at the institution (math)<BR /> 
(6391 NULL out of 7593 rows)
- SATVRMID = Midpoint of SAT scores at the institution (critical reading)<BR /> 
(6400 NULL out of 7593 rows)
- SATWRMID = Midpoint of SAT scores at the institution (writing)<BR /> 
(6860 NULL out of 7593 rows)

COMMENT: I suggest that we do not use any of the columns since there are so much missing data

**ADMISSION AND ENROLLMENT**
- ADM_RATE = Admission rate<BR />
(5422 NULL out of 7593 rows)
- UG = Enrollment of all undergraduate students **(DROPPED)**

COMMENT: One column I already dropepd due to 100% missing data. I suggest that we do not use the other column either since there are so much missing data

**AGE AND GENDER**
- AGE_ENTRY = Average age of entry<BR />
(218 NULL out of 7593 rows)
- UG25ABV = Percentage of undergraduates aged 25 and above<BR />
(805 NULL out of 7593 rows)
- FEMALE = Share of female students<BR />
(218 NULL out of 7593 rows)

COMMENT: Some missing data, but we can work with this. We should go for mean/median/mode to replace. 

**BACKGROUND**
- PAR_ED_PCT_MS = Percent of students whose parents' highest educational level is middle school<BR />
(218 NULL out of 7593 rows)
- PAR_ED_PCT_HS = Percent of students whose parents' highest educational level is high school<BR />
(218 NULL out of 7593 rows)
- PAR_ED_PCT_PS = Percent of students whose parents' highest educational level was is some form of postsecondary education<BR />
(218 NULL out of 7593 rows)

COMMENT: Some missing data, but we can work with this. We should go for mean/median/mode to replace. 

**TRANSFER**
- TRANS_4 = Transfer rate for first-time, full-time students at four-year institutions (within 150% of expected time to completion/6 years)<BR />
(5104 NULL out of 7593 rows)

COMMENT: In my opinion too much missing values.

**COMPLETION**
- C150_4 = Completion rate for first-time, full-time students at four-year institutions (150% of expected time to completion) - **THIS IS OUR Y**<BR />
(5104 NULL out of 7593 rows)

COMMENT: If this is our Y we have a problem. We might want to re-think what we want to look at based on that data we have that is not superbad quality. 

**COSTS AND FEES**
- COSTT4_A = Average cost of attendance (academic year institutions)<BR />
(3673 NULL out of 7593 rows)
- COSTT4_P = Average cost of attendance (program-year institutions)<BR />
(5093 NULL out of 7593 rows)
- INEXPFTE = Instructional expenditures per full-time equivalent student<BR />
(460 NULL out of 7593 rows)
- TUITFTE = Net tuition revenue per full-time equivalent student<BR />
(460 NULL out of 7593 rows)
- TUITIONFEE_IN = In-state tuition and fees<BR />
(3290 NULL out of 7593 rows)
- TUITIONFEE_OUT = Out-of-state tuition and fees<BR />
(3504 NULL out of 7593 rows)

COMMENT: INEXPFTE and TUITFTE we can use, we have to talk about the rest and how to fill the missng values.

**STAFF**
- AVGFACSAL = Average faculty salary<BR />
(3139 NULL out of 7593 rows)
- PFTFAC = Proportion of faculty that is full-time<BR />
(3656 NULL out of 7593 rows)

COMMENT: We need to talk about these as well

**LOANS/DEBTS**
- D_PCTPELL_PCTFLOAN = Number of undergraduate students (denominator percent receiving a pell grant or federal student loan)<BR />
(758 NULL out of 7593 rows)
- PCTFLOAN = Percent of all undergraduate students receiving a federal student loan<BR />
(758 NULL out of 7593 rows)
- GRAD_DEBT_MDN = The median debt for students who have completed<BR />
(29 NULL out of 7593 rows)
- WDRAW_DEBT_MDN = The median debt for students who have not completed<BR />
(29 NULL out of 7593 rows)
- LO_INC_DEBT_MDN = The median debt for students with family income between 0-30,000<BR />
(29 NULL out of 7593 rows)
- MD_INC_DEBT_MDN = The median debt for students with family income between 30,001-75,000<BR />
(29 NULL out of 7593 rows)
- HI_INC_DEBT_MDN = The median debt for students with family income 75,001+<BR />
(29 NULL out of 7593 rows)

COMMENT: Few missing values. Should we us mean/median/mode to replace?

**INCOME**
- MD_EARN_WNE_P6 = Median earnings of students working and not enrolled 6 years after entry **(DROPPED)**

COMMENT: Column dropped due to 100% missing values

**OTHER**
- CONTROL = Control of institution *(I do not understad fromt he documentation what this is)*<BR />
(0 NULL out of 7593 rows)
- LOCALE = Locale of institution *(how populated is the area)*<BR />
(446 NULL out of 7593 rows)
- LOCALE2 = Degree of urbanization of institution **(DROPPED)**
- REGION = Region (IPEDS)<BR />
(0 NULL out of 7593 rows)

COMMENT: We can use the columns that I did not drop. We have to discuss how to handle the missing values of LOCALE. I also need help to understand that CONTROL and LOCAL actually give us for info. We also need to find what number correspond to what kind of region (could not find in documentation).

In [68]:
other['REGION'].unique()

array([5, 3, 8, 6, 4, 7, 1, 2, 0, 9])

In [70]:
big.shape

(7593, 1805)

In [87]:
##Columns with fewer missing values
df = big.dropna(thresh=len(big) - 1000, axis=1)
print(df.columns.values)

['UNITID' 'OPEID' 'OPEID6' 'INSTNM' 'CITY' 'STABBR' 'ZIP' 'ACCREDAGENCY'
 'INSTURL' 'NPCURL' 'HCM2' 'MAIN' 'NUMBRANCH' 'PREDDEG' 'HIGHDEG'
 'CONTROL' 'ST_FIPS' 'REGION' 'LOCALE' 'LATITUDE' 'LONGITUDE' 'CCBASIC'
 'CCUGPROF' 'CCSIZSET' 'HBCU' 'PBI' 'ANNHI' 'TRIBAL' 'AANAPII' 'HSI'
 'NANTI' 'MENONLY' 'WOMENONLY' 'PCIP01' 'PCIP03' 'PCIP04' 'PCIP05'
 'PCIP09' 'PCIP10' 'PCIP11' 'PCIP12' 'PCIP13' 'PCIP14' 'PCIP15' 'PCIP16'
 'PCIP19' 'PCIP22' 'PCIP23' 'PCIP24' 'PCIP25' 'PCIP26' 'PCIP27' 'PCIP29'
 'PCIP30' 'PCIP31' 'PCIP38' 'PCIP39' 'PCIP40' 'PCIP41' 'PCIP42' 'PCIP43'
 'PCIP44' 'PCIP45' 'PCIP46' 'PCIP47' 'PCIP48' 'PCIP49' 'PCIP50' 'PCIP51'
 'PCIP52' 'PCIP54' 'CIP01CERT1' 'CIP01CERT2' 'CIP01ASSOC' 'CIP01CERT4'
 'CIP01BACHL' 'CIP03CERT1' 'CIP03CERT2' 'CIP03ASSOC' 'CIP03CERT4'
 'CIP03BACHL' 'CIP04CERT1' 'CIP04CERT2' 'CIP04ASSOC' 'CIP04CERT4'
 'CIP04BACHL' 'CIP05CERT1' 'CIP05CERT2' 'CIP05ASSOC' 'CIP05CERT4'
 'CIP05BACHL' 'CIP09CERT1' 'CIP09CERT2' 'CIP09ASSOC' 'CIP09CERT4'
 'CIP09BACHL' 'CIP10CERT1'

In [88]:
df.shape

(7593, 408)